sklearn/datasets/tests/test_base.py

```diff
@@ -391,3 +396,263 @@ def test_fetch_remote_raise_warnings_with_invalid_url(monkeypatch):
+
+
+def test_fetch_file_without_sha256(monkeypatch, tmpdir):
+    server_side = tmpdir.mkdir("server_side")
+    data_file = Path(server_side / "data.jsonl")
+    server_data = '{"a": 1, "b": 2}\n'
+    data_file.write_text(server_data, encoding="utf-8")
+
+    client_side = tmpdir.mkdir("client_side")
+
+    urlretrieve_mock = _mock_urlretrieve(server_side)
+    monkeypatch.setattr("sklearn.datasets._base.urlretrieve", urlretrieve_mock)
+
+    # The first call should trigger a download:
+    fetched_file_path = fetch_file(
+        "https://example.com/data.jsonl",
+        folder=client_side,
+    )
+    assert fetched_file_path == client_side / "data.jsonl"
+    assert fetched_file_path.read_text(encoding="utf-8") == server_data
+    assert urlretrieve_mock.call_count == 1
+
+    # Fetching again the same file to the same folder should do nothing:
+    fetched_file_path = fetch_file(
+        "https://example.com/data.jsonl",
+        folder=client_side,
+    )
+    assert fetched_file_path == client_side / "data.jsonl"
+    assert fetched_file_path.read_text(encoding="utf-8") == server_data
+    assert urlretrieve_mock.call_count == 1
+
+    # Deleting and calling again should re-download
+    fetched_file_path.unlink()
+    fetched_file_path = fetch_file(
+        "https://example.com/data.jsonl",
+        folder=client_side,
+    )
+    assert fetched_file_path == client_side / "data.jsonl"
+    assert fetched_file_path.read_text(encoding="utf-8") == server_data
+    assert urlretrieve_mock.call_count == 2
+
+
+def test_fetch_file_with_sha256(monkeypatch, tmpdir):
+    server_side = tmpdir.mkdir("server_side")
+    data_file = Path(server_side / "data.jsonl")
+    server_data = '{"a": 1, "b": 2}\n'
+    data_file.write_text(server_data, encoding="utf-8")
+    expected_sha256 = hashlib.sha256(data_file.read_bytes()).hexdigest()
+
+    client_side = tmpdir.mkdir("client_side")
+
+    urlretrieve_mock = _mock_urlretrieve(server_side)
+    monkeypatch.setattr("sklearn.datasets._base.urlretrieve", urlretrieve_mock)
+
+    # The first call should trigger a download.
+    fetched_file_path = fetch_file(
+        "https://example.com/data.jsonl", folder=client_side, sha256=expected_sha256
+    )
+    assert fetched_file_path == client_side / "data.jsonl"
+    assert fetched_file_path.read_text(encoding="utf-8") == server_data
+    assert urlretrieve_mock.call_count == 1
+
+    # Fetching again the same file to the same folder should do nothing when
+    # the sha256 match:
+    fetched_file_path = fetch_file(
+        "https://example.com/data.jsonl", folder=client_side, sha256=expected_sha256
+    )
+    assert fetched_file_path == client_side / "data.jsonl"
+    assert fetched_file_path.read_text(encoding="utf-8") == server_data
+    assert urlretrieve_mock.call_count == 1
+
+    # Corrupting the local data should yield a warning and trigger a new download:
+    fetched_file_path.write_text("corruped contents", encoding="utf-8")
```

Comment:
I was looking for the test of the local file with wrong sha256 edge-case, glad that you have added it!

Not that it matters much for the test but there is a typo :wink:
```suggestion
    fetched_file_path.write_text("corrupted contents", encoding="utf-8")
```